betty/extension/_webpack/__init__.py

```diff
@@ -0,0 +1,291 @@
+        webpack_entry = {}
+        for entrypoint_provider in entrypoint_providers:
+            extension_webpack_entrypoint_directory_path = (
+                entrypoint_provider.webpack_entrypoint_directory_path()
+            )
+            extension_working_directory_path = (
+                entrypoints_working_directory_path / entrypoint_provider.name()
+            )
+            await to_thread(
+                copytree,
+                extension_webpack_entrypoint_directory_path,
+                extension_working_directory_path,
+            )
+            working_package_json_dependencies[entrypoint_provider.name()] = (
+                f"file:{extension_working_directory_path}"
+            )
+            webpack_entry[entrypoint_provider.name()] = str(
+                # Use os.path because pathlib has trouble with leading dots.
+                join(
+                    ".",
+                    (extension_working_directory_path / "main.ts").relative_to(
+                        working_directory_path
+                    ),
+                )
+            )
+        webpack_configuration_json = dumps(
+            {
+                "debug": self._app.project.configuration.debug,
+                "cacheDirectory": str(
+                    self._app.binary_file_cache.with_scope("webpack").path
+                ),
+                "entry": webpack_entry,
+            }
+        )
+        async with aiofiles.open(
+            working_directory_path / "webpack.config.json", "w"
+        ) as configuration_f:
+            await configuration_f.write(webpack_configuration_json)
+        await gather(
+            *[
+                self._app.renderer.render_file(
+                    file_path,
+                    job_context=job_context,
+                )
+                async for file_path in iterfiles(working_directory_path)
+            ]
+        )
+
+        # Add dependencies to package.json.
+        working_package_json_path = working_directory_path / "package.json"
+        async with aiofiles.open(
+            working_package_json_path, "r"
+        ) as working_package_json_f:
+            working_package_json = loads(await working_package_json_f.read())
+        working_package_json["dependencies"].update(working_package_json_dependencies)
+        async with aiofiles.open(
+            working_package_json_path, "w"
+        ) as working_package_json_f:
+            await working_package_json_f.write(dumps(working_package_json))
+
+        # Install third-party dependencies.
+        install_arguments = ["install"]
+        if not self._app.project.configuration.debug:
+            install_arguments.append("--production")
+        await npm(install_arguments, cwd=working_directory_path)
+
+        # Run Webpack.
+        await npm(("run", "webpack"), cwd=working_directory_path)
+
+        build_directory_path = working_directory_path / "build"
+
+        # Ensure there is always a vendor.css
```

Comment:
"... even if no endpoints provide any, so pages can unconditionally link the CSS file"